src/Test/StrictCheck/Observe.hs

```diff
@@ -50,17 +54,14 @@ import Test.StrictCheck.Demand
 -- This tells us that our context did indeed evaluate the result of @reverse@
 -- to force only its first constructor, and that doing so required the entire
 -- spine of the list to be evaluated, but did not evaluate any of its elements.
-{-# NOINLINE observe1 #-}
```

Comment:
We still need a `{-# NOINLINE #-}` annotation surrounding `observe1` and `observeNP`, to prevent the `unsafePerformIO` from being lifted out to the top level by CSE optimization.

Reply:
Do we? If all the inputs to two calls of `observe*` are the same the output will be the same and it does not matter if we do not run the `IO` action twice. However if the inputs differ, then it will run the whole `IO` action again, ensuring that we entangle the inputs again, thus give the correct answer. Previously the problem was, that CSE would memorize the entangled inputs even when parts of the other inputs to `observe*` were differed. But that shouldn't be possible anymore or am I missing something? My understanding is that calling `unsafePerformIO` on an `IO` action that always returns the same result and has no side effects is safe.

Reply:
CSE can't make it so that only part of the action under `unsafePerformIO` is executed. And executing the whole action will always result in the correct answer.